io/bmp/ReadBmp.ml

```diff
@@ -0,0 +1,169 @@
+open Bigarray
+
+type bicompression = BI_RGB | BI_RLE8 | BI_RLE4 | BI_BITFIELDS
+type bibitcount = Monochrome | Color16 | Color256 | ColorRGB | ColorRGBA
+
+type bitmapfileheader = {
+  bfType : int;
+  bfSize : int;
+  bfReserved1 : int;
+  bfReserved2 : int;
+  bfOffBits : int;
+}
+
+type bitmapinfoheader = {
+  biSize : int;
+  biWidth : int;
+  biHeight : int;
+  biPlanes : int;
+  biBitCount : bibitcount;
+  biCompression : bicompression;
+  biSizeImage : int;
+  biXPelsPerMeter : int;
+  biYPelsPerMeter : int;
+  biClrUsed : int;
+  biClrImportant : int;
+}
+
+type bmp = {
+  bmpFileHeader : bitmapfileheader;
+  bmpInfoHeader : bitmapinfoheader;
+  bmpBytes : (int32, Bigarray.int32_elt, Bigarray.c_layout) Bigarray.Array1.t;
+}
+
+let bytes_read = ref 0
+
+let read_byte ic =
+  incr bytes_read;
+  input_byte ic
+
+let skip_byte ic =
+  incr bytes_read;
+  ignore (input_byte ic)
+
+let read_16bit ic =
+  let b0 = read_byte ic in
+  let b1 = read_byte ic in
+  (b1 lsl 8) + b0
+
+let read_32bit ic =
+  let b0 = read_byte ic in
+  let b1 = read_byte ic in
+  let b2 = read_byte ic in
+  let b3 = read_byte ic in
+  (b3 lsl 24) + (b2 lsl 16) + (b1 lsl 8) + b0
+
+let read_bit_count ic =
+  match read_16bit ic with
+  | 1 -> Monochrome
+  | 4 -> Color16
+  | 8 -> Color256
+  | 24 -> ColorRGB
+  | 32 -> ColorRGBA
+  | n -> failwith ("invalid number of colors in bitmap: " ^ string_of_int n)
+
+let read_compression ic =
+  match read_32bit ic with
+  | 0 -> BI_RGB
+  | 1 -> BI_RLE8
+  | 2 -> BI_RLE4
+  | 3 -> BI_BITFIELDS
+  | n -> failwith ("invalid compression: " ^ string_of_int n)
+
+let load_bitmapfileheader ic =
+  let bfType = read_16bit ic in
+  if bfType <> 19778 then failwith "Invalid bitmap file";
+  let bfSize = read_32bit ic in
+  let bfReserved1 = read_16bit ic in
+  let bfReserved2 = read_16bit ic in
+  let bfOffBits = read_32bit ic in
+  { bfType; bfSize; bfReserved1; bfReserved2; bfOffBits }
+
+let load_bitmapinfoheader ic =
+  try
+    let biSize = read_32bit ic in
```

Comment:
This is only cosmetic but you can use `let ... and ... in ...` here because all the bindings are independent of each other:

```ocaml
let biSize = read_32bit ic
and biWidth = read_32bit ic
and ... in
{
  biSize;
  biWidth;
  ...
}
```
```

Reply:
I do prefere avoid and when possible in ocaml